File: src/lava/magma/runtime/message_infrastructure/test/test_channel.py
# Copyright (C) 2022 Intel Corporation
# SPDX-License-Identifier: BSD-3-Clause
# See: https://spdx.org/licenses/

import numpy as np
import unittest
from functools import partial
import time

from message_infrastructure.multiprocessing import MultiProcessing

from message_infrastructure import (
    ChannelBackend,
    Channel,
    SendPort,
    RecvPort,
    SupportGRPCChannel,
    ChannelQueueSize
)


def prepare_data():
    return np.random.random_sample((2, 4))


const_data = prepare_data()


def actor_stop(name):
    pass


def send_proc(*args, **kwargs):
    actor = args[0]
    actor.set_stop_fn(partial(actor_stop, "send"))
    port = kwargs.pop("port")
    if not isinstance(port, SendPort):
        raise AssertionError()
    port.start()
    port.send(const_data)
    port.join()
    actor.status_stopped()


def recv_proc(*args, **kwargs):
    actor = args[0]
    actor.set_stop_fn(partial(actor_stop, "recv"))
    port = kwargs.pop("port")
    port.start()
    if not isinstance(port, RecvPort):
        raise AssertionError()
    data = port.recv()
    if not np.array_equal(data, const_data):
        raise AssertionError()
    port.join()
    actor.status_stopped()


class Builder:
    def build(self):
        pass


class TestChannel(unittest.TestCase):

    def test_shmemchannel(self):
        mp = MultiProcessing()
        mp.start()
        nbytes = np.prod(const_data.shape) * const_data.dtype.itemsize
        name = 'test_shmem_channel'

        shmem_channel = Channel(
            ChannelBackend.SHMEMCHANNEL,
            ChannelQueueSize,
            nbytes,
            name,
            name)

        send_port = shmem_channel.src_port
        recv_port = shmem_channel.dst_port

        recv_port_fn = partial(recv_proc, port=recv_port)
        send_port_fn = partial(send_proc, port=send_port)

        builder1 = Builder()
        builder2 = Builder()
        mp.build_actor(recv_port_fn, builder1)
        mp.build_actor(send_port_fn, builder2)

        time.sleep(0.1)
        mp.stop(True)

    def test_single_process_shmemchannel(self):
        predata = prepare_data()
        nbytes = np.prod(predata.shape) * predata.dtype.itemsize
        name = 'test_single_process_shmem_channel'

        shmem_channel = Channel(
            ChannelBackend.SHMEMCHANNEL,
            ChannelQueueSize,
            nbytes,
            name,
            name)

        send_port = shmem_channel.src_port
        recv_port = shmem_channel.dst_port

        send_port.start()
        recv_port.start()

        send_port.send(predata)
        resdata = recv_port.recv()

        if not np.array_equal(resdata, predata):
            raise AssertionError()

        send_port.join()
        recv_port.join()

    def test_socketchannel(self):
        mp = MultiProcessing()
        mp.start()
        nbytes = np.prod(const_data.shape) * const_data.dtype.itemsize
        name = 'test_socket_channel'

        socket_channel = Channel(
            ChannelBackend.SOCKETCHANNEL,
            ChannelQueueSize,
            nbytes,
            name,
            name)

        send_port = socket_channel.src_port
        recv_port = socket_channel.dst_port

        recv_port_fn = partial(recv_proc, port=recv_port)
        send_port_fn = partial(send_proc, port=send_port)

        builder1 = Builder()
        builder2 = Builder()
        mp.build_actor(recv_port_fn, builder1)
        mp.build_actor(send_port_fn, builder2)

        time.sleep(0.1)
        mp.stop(True)

    def test_single_process_socketchannel(self):
        predata = prepare_data()
        nbytes = np.prod(predata.shape) * predata.dtype.itemsize
        name = 'test_single_process_socket_channel'

        socket_channel = Channel(
            ChannelBackend.SOCKETCHANNEL,
            ChannelQueueSize,
            nbytes,
            name,
            name)

        send_port = socket_channel.src_port
        recv_port = socket_channel.dst_port

        send_port.start()
        recv_port.start()

        send_port.send(predata)
        resdata = recv_port.recv()

        if not np.array_equal(resdata, predata):
            raise AssertionError()

        send_port.join()
        recv_port.join()

    @unittest.skipIf(not SupportGRPCChannel, "Not support grpc channel.")
    def test_grpcchannel(self):
        from message_infrastructure import GetRPCChannel
        mp = MultiProcessing()
        mp.start()
        name = 'test_grpc_channel'
        url = '127.13.2.11'
        port = 8003
        grpc_channel = GetRPCChannel(
            url,
            port,
            name,
            name,
            ChannelQueueSize)

        send_port = grpc_channel.src_port
        recv_port = grpc_channel.dst_port

        recv_port_fn = partial(recv_proc, port=recv_port)
        send_port_fn = partial(send_proc, port=send_port)

        builder1 = Builder()
        builder2 = Builder()
        mp.build_actor(recv_port_fn, builder1)
        mp.build_actor(send_port_fn, builder2)

        time.sleep(0.1)
        mp.stop(True)


if __name__ == "__main__":
    unittest.main()

File: src/lava/magma/runtime/message_infrastructure/test/test_channel.cc
// Copyright (C) 2022 Intel Corporation
// SPDX-License-Identifier: BSD-3-Clause
// See: https://spdx.org/licenses/

#include <iostream>

#include <gtest/gtest.h>
#include <multiprocessing.h>
#include <abstract_actor.h>
#include <shmem_channel.h>
#include <channel_proxy.h>

#include <pybind11/pybind11.h>
#include <pybind11/numpy.h>
namespace py = pybind11;

using namespace message_infrastructure;

class Builder {
  public:
    void Build() {};
};

void TargetFunction(Builder builder, AbstractActor* actor_ptr) {
  std::cout << "Target Function running..." << std::endl;
  actor_ptr->SetStatus(ActorStatus::StatusStopped);
  builder.Build();
}

void ActorStop(std::string actor_name) {
  std::cout << actor_name << " STOP" << std::endl;
}

int* DataInteger() {
  int data = 42;
  int *data_ptr;
  data_ptr = &data;
  return data_ptr;
}

void SendProc(ShmemSendPort send_port, MetaDataPtr data) {
  // Sends data
  send_port.Start();
  send_port.Send(data);
  send_port.Join();
}

auto RecvProc(ShmemRecvPort recv_port) {
  // Returns received data
  recv_port.Start();
  auto recv_data = recv_port.Recv();
  recv_port.Join();

  return recv_data;
}

TEST(TestSharedMemory, SharedMemSendReceive) {
  // Creates a pair of send and receive ports
  // TODO: Define success criteria
  MultiProcessing mp;
  Builder *builder = new Builder();

  AbstractActor::TargetFn target_fn;

  auto bound_fn = std::bind(&TargetFunction, (*builder),  std::placeholders::_1);
  target_fn = bound_fn;

  AbstractActor::StopFn stop_fn;
  AbstractActor::StopFn stop_fn2;

  auto actor_stop_fn = std::bind(&ActorStop, std::placeholders::_1);
  stop_fn = actor_stop_fn("Send");
  
  AbstractActor* send_actor_ptr;
  send_actor_ptr->SetStopFn(stop_fn);
  send_actor_ptr->SetStatus(ActorStatus::StatusStopped);

  AbstractActor* recv_actor_ptr;
  recv_actor_ptr->SetStopFn(stop_fn2);
  recv_actor_ptr->SetStatus(ActorStatus::StatusStopped);

  int size = 1;
  int nbytes = sizeof(int);
  std::string name = "test_shmem_channel";
  std::string src_name = "Source1";
  std::string dst_name = "Dest1";
  auto shmem_channel = ShmemChannel(
    src_name,
    dst_name,
    size,
    nbytes);

  // AbstractSendPortPtr send_port = *shmem_channel.GetSendPort();
  // AbstractRecvPortPtr recv_port = *shmem_channel->GetRecvPort();

  // auto send_port_fn = std::bind(&SendProc, send_port);
  // auto send_port_fn = std::bind(&SendProc, (*builder_send), SendPort);
  // auto recv_port_fn = std::bind(&RecvProc, (*builder_recv), RecvPort);
  // mp.BuildActor(send_port_fn);
  // mp.BuildActor(recv_port_fn);
  
  // Stop any currently running actors
  // mp.Stop(true);
}

TEST(TestSharedMemory, SharedMemSingleProcess){
  ChannelProxy ShmemChannel;

  SendPortProxyPtr SendPort = ShmemChannel.GetSendPort();
  RecvPortProxyPtr RecvPort = ShmemChannel.GetRecvPort();

  SendPort.Start();
  RecvPort.Start();
  auto data = DataInteger();
  SendPort.Send(data);
  auto received_data = RecvPort.Recv();
  EXPECT_EQ(data, received_data)
}
